storage/range.go

```diff
@@ -278,20 +279,76 @@ func (r *Range) IsFirstRange() bool {
 	return bytes.Equal(r.Desc().StartKey, engine.KeyMin)
 }
 
-// IsLeader returns true if this range replica is the raft leader.
-// TODO(spencer): this is always true for now.
-func (r *Range) IsLeader() bool {
-	return true
+// getLease returns the current leader lease.
+func (r *Range) getLease() *proto.Lease {
+	return (*proto.Lease)(atomic.LoadPointer(&r.lease))
+}
+
+// HasLeaderLease returns true if this range replica currently holds
+// the Raft leader lease. It's possible that the replica is not in
+// fact the leader, but holding the lease means that no other replica
```

Comment:
I should mention that the only reason to check the lease when doing writes is to avoid the raft failure (and its possibly high latency) when we're already reasonably certain that we're not the leader. We may want to relax this though to be more specific. Only return a NotLeaderError for writes immediately if we're SURE we can't be the leader--that is, another replica currently has -or- had the lease last.